### PR TITLE
Implement decaying D8 flow accumulation

### LINKED ISSUE #386: Contribute Pollutant Decay to Flow Accumulation
For the NCI Water Quality work, I implemented a modification to D8 Flow Accumulation that allowed for the on-pixel load (`1` per pixel, or else the pixel value of a weights raster, if used) to decay by some known proportion as it traveled downstream.  Rafa said there could be some applications in InVEST for modeling pollutant decay in this way.

For example, suppose we have the following flowpath, weights and decay factors:

```
Flowpath: a --> b --> c --> d
Weights:  1     2     3     4
Decay:    0.5   0.75  0.3   0.1
```

On pixel `a`, the flow accumulation would be `1`.

On pixel `b`, the flow accumulation would be the decayed contribution from upstream, plus the on-pixel load from `b`, `1*0.5**1 + 2*0.75**0`.

On pixel `c`, the flow accumulation would follow the same pattern, `1*0.5**2 + 2*0.75**1 + 3*0.3**0`

Pixel `d` would again follow the same pattern: `1*0.5**3 + 2*0.75**2 + 3*0.3**1 + 4*0.1**0`.

That's the idea anyways, at least for a simple flowpath.  For real flowpaths, the best way to do this is actually to track the upstream contributions so that we can sum them correctly in cases of convergent flow and non-linear flowpaths.

### PR DESCRIPTION
**Pull request by phargogh**
This PR copies the modifications to D8 flow accumulation implemented for the NCI Water Quality project that allow for pixel loads to decay as they pass over the landscape.  You could think of like a pollutant flowing across the landscape, where some fraction is absorbed due to landcover or some other factor.

Tests have been implemented and runtime should only trivially be affected when decaying accumulation is not in use.  When decaying accumulation is used, there is an extra cost in memory and in iteration over the pollutant contributions to the pixel.

Fixes #386

### CODE REVIEW

File: src/pygeoprocessing/routing/routing.pyx [hunk 1]
```diff
@@ -138,6 +138,18 @@ cdef struct FlowPixelType:
     int last_flow_dir
     double value
 
+cdef struct DecayingValue:
+    double decayed_value  # The value, which will be progressively updated as it decays
+    double min_value  # The minimum value before the Decaying Value should be ignored.
+
+cdef struct WeightedFlowPixelType:
```
Comment (dcdenu4) [inline src/pygeoprocessing/routing/routing.pyx]: I noticed the other `struct` definitions have a few comments above to describe them. Should these follow that same convention?

File: src/pygeoprocessing/routing/routing.pyx [hunk 2]
```diff
@@ -138,6 +138,18 @@ cdef struct FlowPixelType:
...
     double value
 
+cdef struct DecayingValue:
+    double decayed_value  # The value, which will be progressively updated as it decays
+    double min_value  # The minimum value before the Decaying Value should be ignored.
+

+cdef struct WeightedFlowPixelType:
+    int xi
+    int yi
```
Comment (dcdenu4) [inline src/pygeoprocessing/routing/routing.pyx]: The other `struct` definitions use `unsigned int` for pixel location, any reason not to do that here as well for consistency?

File: src/pygeoprocessing/routing/routing.pyx [hunk 3]
```diff
@@ -1593,7 +1643,9 @@ def flow_accumulation_d8(
         xi_n = -1
         yi_n = -1
 
-        # search block for to set flow direction
+        # Search the block for root pixels.
+        # A root pixel is a flow direction pixel that is nodata, which means it
+        # may be a drain just off the edge.
```
Comment (dcdenu4) [inline src/pygeoprocessing/routing/routing.pyx:1654]: Thank you!

File: src/pygeoprocessing/routing/routing.pyx (hunk 1 above)
Comment (dcdenu4) [inline src/pygeoprocessing/routing/routing.pyx:152]: I think that this replaces ALL instances of `FlowPixelType`. Should we remove that struct definition?

**Review by dcdenu4 — CHANGES_REQUESTED** [review]
Hey @phargogh, sorry for the LONG delay here. This looks solid to me, I just had a few questions related to consistency.

File: src/pygeoprocessing/routing/routing.pyx (hunk 1 above)
Comment (phargogh) [reply src/pygeoprocessing/routing/routing.pyx]: Yep, great catch!  Patched in 4f0e062d

Comment (phargogh) [reply src/pygeoprocessing/routing/routing.pyx:152]: Interestingly, I'm still seeing `FlowPixelType` used in `flow_accumulation_mfd`, so I think it makes sense to leave it here for now.

File: src/pygeoprocessing/routing/routing.pyx (hunk 2 above)
Comment (phargogh) [reply src/pygeoprocessing/routing/routing.pyx]: Nope!  In fact they really should be `unsigned int`s for consistency with the way indexes are used elsewhere in the function.  Patched in d645b70a

**Review by dcdenu4 — APPROVED** [review]
Thanks!

**Pull request merged**